# Napier's Logarithms: 九九のない世界

* [John Napier - Wikipedia](https://en.wikipedia.org/wiki/John_Napier)
  * [Napierian logarithm - Wikipedia](https://en.wikipedia.org/wiki/Napierian_logarithm#:~:text=The%20term%20Napierian%20logarithm%20or,it%20is%20named%20after%20him.)
  * [Logarithms: The Early History of a Familiar Function - John Napier Introduces Logarithms | Mathematical Association of America](https://www.maa.org/press/periodicals/convergence/logarithms-the-early-history-of-a-familiar-function-john-napier-introduces-logarithms)

Napierは、$N$に対して対数 (logarithm) $L$を次の数式で定義し、$R=7$ の場合の数表を作成した。

$$
  \frac{N}{10^R} = \left(1-\frac{1}{10^{R}}\right)^{L}
$$

* [e (mathematical constant) - Wikipedia](https://en.wikipedia.org/wiki/E_%28mathematical_constant%29)

その後、Eulerは自然対数$e$を次の極限式で定義した。

$$
  \lim_{n\rightarrow\infty} \left( 1+\frac{1}{n} \right)^n = e
$$

$$
  \lim_{n\rightarrow\infty} \left( 1-\frac{1}{n} \right)^n = \frac{1}{e}
$$

$R$が十分大きいとき次の近似式が得られる。

$$
\left( 1-\frac{1}{10^R} \right)^{10^R} = (1-10^{-R})^{10^R} \approx \frac{1}{e}
$$

In [4]:
import math
(1-10**-5)**(10**5), (1-10**-7)**(10**7), 1/math.e

(0.3678776017682465, 0.367879422971105, 0.36787944117144233)

現代数学の$\log$を用いるとNapierの対数$L$は次のように近似できる。

$$
  L = \log_{1-10^{-R}} \frac{N}{10^R} \approx
  10^R \log_{1/e}\left( \frac{N}{10^R} \right) \approx -10^R \log_{e}\left( \frac{N}{10^R} \right)
$$

## Napierの定義

In [7]:
1-10**-5, 1-10**-7

(0.99999, 0.9999999)

$p_{n}$ を次の数列として定義したとき、

$$
\begin{eqnarray}
p_{0}   &=& 10^7 = 10000000 \\
p_{1}   &=& 10^7 \left(1-\frac{1}{10^7} \right)= 10000000 \times 0.9999999 = 9999999 \\
p_{n+1} &=& p_{n} \left(1-\frac{1}{10^7}\right) = p_{n} \times 0.9999999 
\end{eqnarray}
$$

数列の添字 $n$ は $p_{n}$ のNapier対数である。

$$
p_{n} = 10^7 \times \left(1-\frac{1}{10^7}\right)^n
$$

$$
  n = \log_{1-10^{-7}}\left(\frac{p_{n}}{10^7}\right)
$$

次のように対数法則を満たす。

$$
p_{m+n} = 10^R \times \left(1-\frac{1}{10^7}\right)^{m+n} = 10^R \times \left(1-\frac{1}{10^7}\right)^m \times \left(1-\frac{1}{10^7}\right)^n = p_{m} \times p_{n} \times \frac{1}{10^R}
$$

### $R=5$のとき

In [61]:
import math

def napier5(n):
    """
    R=5のときのNapierの数表の第n項を計算する
    """
    return 10**5 * (1-1/(10**5))**n

In [76]:
# `math.log()`を使って、Napierの対数を計算する
list(map(lambda x: math.log(x/10**5, 1-1/(10**5)), (1234, 5678)))

[439488.72859419696, 286855.6785070823]

In [77]:
# Napierの数表の項番を計算する
list(map(napier5, _))

[1234.0, 5677.999999999997]

In [78]:
# 対数法則から積を求める
napier5(sum(__))*10**5

7006651.999999997

In [85]:
# 検算
1234*5678

7006652

:::{seealso}
* [Compound interest - Wikipedia](https://en.wikipedia.org/wiki/Compound_interest)
    - (1655&mdash;1705) [Jacob Bernoulli - Wikipedia](https://en.wikipedia.org/wiki/Jacob_Bernoulli)
:::

### 出版された表を再現する

Napierは、$R=7$の場合の数表を作成した。

In [20]:
!pip install -U astropy

  Using cached astropy-6.1.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)
  Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.7 kB)
Using cached astropy-6.1.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 36.8 MB/s eta 0:00:0000:01
Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (738 kB)


In [21]:
from astropy.coordinates import Angle
from astropy import units

In [22]:
Angle('{:d}°{:d}′{:d}″'.format(1,2,3)).radian

0.018049613347708025

In [23]:
import math
import numpy as np
#from scoping import scoping
import pandas as pd

In [24]:
def napier(dms_d=0, r=7):
    for dms_m in range(60):
        a = Angle('{:d}°{:d}′{:d}″'.format(dms_d,dms_m,0))
        yield (a,
               "{:8.0f}".format(math.sin(a.radian)*10**r),
               "{:8.0f}".format(math.log(math.sin(a.radian), 1-1/(10**r)) if a.radian else float('inf')),
               "{:8.0f}".format(math.log(math.tan(a.radian), 1-1/(10**r)) if a.radian else float('inf')),
               "{:8.0f}".format(math.log(math.cos(a.radian), 1-1/(10**r))),
               "{:8.0f}".format(math.cos(a.radian)*10**r),
               Angle('{:d}°{:d}′{:d}″'.format(90,0,0))-Angle('{:d}°{:d}′{:d}″'.format(dms_d,dms_m,0)))

```{note}
角度が$0\degree$から$1\degree$の正弦、余弦とそれらの対数、および正接の対数の値。角度は一分刻み
```

In [25]:
pd.DataFrame(napier(dms_d=0), columns=['d','sin','log(sin(d))','log(tan(d))','log(cos(d))', 'cos(d)', '90-d'])

d       sin log(sin(d)) log(tan(d)) log(cos(d))    cos(d)  \
0   0d00m00s         0         inf         inf          -0  10000000   
1   0d01m00s      2909    81425711    81425711           0  10000000   
2   0d02m00s      5818    74494240    74494239           2   9999998   
3   0d03m00s      8727    70439590    70439586           4   9999996   
4   0d04m00s     11636    67562771    67562764           7   9999993   
5   0d05m00s     14544    65331336    65331326          11   9999989   
6   0d06m00s     17453    63508123    63508107          15   9999985   
7   0d07m00s     20362    61966618    61966597          21   9999979   
8   0d08m00s     23271    60631306    60631279          27   9999973   
9   0d09m00s     26180    59453478    59453444          34   9999966   
10  0d10m00s     29089    58399876    58399833          42   9999958   
11  0d11m00s     31998    57446777    57446726          51   9999949   
12  0d12m00s     34907    56576666    56576605          61   9999939   
13  0d13m00s     37815    55776243    55776171          72   9999928   
14  0d14m00s     40724    55035167    55035084          83   9999917   
15  0d15m00s     43633    54345242    54345147          95   9999905   
16  0d16m00s     46542    53699861    53699753         108   9999892   
17  0d17m00s     49451    53093620    53093498         122   9999878   
18  0d18m00s     52360    52522041    52521904         137   9999863   
19  0d19m00s     55268    51981374    51981221         153   9999847   
20  0d20m00s     58177    51468446    51468277         169   9999831   
21  0d21m00s     61086    50980551    50980364         187   9999813   
22  0d22m00s     63995    50515356    50515152         205   9999795   
23  0d23m00s     66904    50070845    50070621         224   9999776   
24  0d24m00s     69813    49645256    49645012         244   9999756   
25  0d25m00s     72721    49237043    49236778         264   9999736   
26  0d26m00s     75630    48844843    48844557         286   9999714   
27  0d27m00s     78539    48467447    48467139         308   9999692   
28  0d28m00s     81448    48103778    48103447         332   9999668   
29  0d29m00s     84357    47752873    47752517         356   9999644   
30  0d30m00s     87265    47413866    47413485         381   9999619   
31  0d31m00s     90174    47085976    47085570         407   9999593   
32  0d32m00s     93083    46768498    46768065         433   9999567   
33  0d33m00s     95992    46460791    46460330         461   9999539   
34  0d34m00s     98900    46162271    46161782         489   9999511   
35  0d35m00s    101809    45872405    45871887         518   9999482   
36  0d36m00s    104718    45590706    45590158         548   9999452   
37  0d37m00s    107627    45316727    45316148         579   9999421   
38  0d38m00s    110535    45050055    45049444         611   9999389   
39  0d39m00s    113444    44790311    44789668         644   9999357   
40  0d40m00s    116353    44537144    44536467         677   9999323   
41  0d41m00s    119261    44290229    44289518         711   9999289   
42  0d42m00s    122170    44049266    44048519         746   9999254   
43  0d43m00s    125079    43813973    43813190         782   9999218   
44  0d44m00s    127987    43584090    43583271         819   9999181   
45  0d45m00s    130896    43359374    43358517         857   9999143   
46  0d46m00s    133805    43139598    43138702         895   9999105   
47  0d47m00s    136713    42924549    42923614         935   9999065   
48  0d48m00s    139622    42714028    42713053         975   9999025   
49  0d49m00s    142530    42507849    42506833        1016   9998984   
50  0d50m00s    145439    42305836    42304778        1058   9998942   
51  0d51m00s    148348    42107824    42106723        1100   9998900   
52  0d52m00s    151256    41913657    41912513        1144   9998856   
53  0d53m00s    154165    41723190    41722002        1188   9998812   
54  0d54m00s    157073    41536284    41535050     

```{note}
角度が$89\degree$から$90\degree$の正弦、余弦とそれらの対数、および正接の対数の値。角度は一分刻み
```

In [26]:
pd.DataFrame(napier(dms_d=89), columns=['d','sin','log(sin(d))','log(tan(d))','log(cos(d))', 'cos(d)', '90-d'])

d       sin log(sin(d)) log(tan(d)) log(cos(d))    cos(d)  \
0   89d00m00s   9998477        1523   -40481252    40482775    174524   
1   89d01m00s   9998527        1473   -40649357    40650830    171616   
2   89d02m00s   9998577        1423   -40820334    40821758    168707   
3   89d03m00s   9998625        1375   -40994284    40995659    165799   
4   89d04m00s   9998673        1327   -41171312    41172639    162890   
5   89d05m00s   9998720        1280   -41351528    41352808    159982   
6   89d06m00s   9998766        1234   -41535050    41536284    157073   
7   89d07m00s   9998812        1188   -41722002    41723190    154165   
8   89d08m00s   9998856        1144   -41912513    41913657    151256   
9   89d09m00s   9998900        1100   -42106723    42107824    148348   
10  89d10m00s   9998942        1058   -42304778    42305836    145439   
11  89d11m00s   9998984        1016   -42506833    42507849    142530   
12  89d12m00s   9999025         975   -42713053    42714028    139622   
13  89d13m00s   9999065         935   -42923614    42924549    136713   
14  89d14m00s   9999105         895   -43138702    43139598    133805   
15  89d15m00s   9999143         857   -43358517    43359374    130896   
16  89d16m00s   9999181         819   -43583271    43584090    127987   
17  89d17m00s   9999218         782   -43813190    43813973    125079   
18  89d18m00s   9999254         746   -44048519    44049266    122170   
19  89d19m00s   9999289         711   -44289518    44290229    119261   
20  89d20m00s   9999323         677   -44536467    44537144    116353   
21  89d21m00s   9999357         644   -44789668    44790311    113444   
22  89d22m00s   9999389         611   -45049444    45050055    110535   
23  89d23m00s   9999421         579   -45316148    45316727    107627   
24  89d24m00s   9999452         548   -45590158    45590706    104718   
25  89d25m00s   9999482         518   -45871887    45872405    101809   
26  89d26m00s   9999511         489   -46161782    46162271     98900   
27  89d27m00s   9999539         461   -46460330    46460791     95992   
28  89d28m00s   9999567         433   -46768065    46768498     93083   
29  89d29m00s   9999593         407   -47085570    47085976     90174   
30  89d30m00s   9999619         381   -47413485    47413866     87265   
31  89d31m00s   9999644         356   -47752517    47752873     84357   
32  89d32m00s   9999668         332   -48103447    48103778     81448   
33  89d33m00s   9999692         308   -48467139    48467447     78539   
34  89d34m00s   9999714         286   -48844557    48844843     75630   
35  89d35m00s   9999736         264   -49236778    49237043     72721   
36  89d36m00s   9999756         244   -49645012    49645256     69813   
37  89d37m00s   9999776         224   -50070621    50070845     66904   
38  89d38m00s   9999795         205   -50515152    50515356     63995   
39  89d39m00s   9999813         187   -50980364    50980551     61086   
40  89d40m00s   9999831         169   -51468277    51468446     58177   
41  89d41m00s   9999847         153   -51981221    51981374     55268   
42  89d42m00s   9999863         137   -52521904    52522041     52360   
43  89d43m00s   9999878         122   -53093498    53093620     49451   
44  89d44m00s   9999892         108   -53699753    53699861     46542   
45  89d45m00s   9999905          95   -54345147    54345242     43633   
46  89d46m00s   9999917          83   -55035084    55035167     40724   
47  89d47m00s   9999928          72   -55776171    55776243     37815   
48  89d48m00s   9999939          61   -56576605    56576666     34907   
49  89d49m00s   9999949          51   -57446726    57446777     31998   
50  89d50m00s   9999958          42   -58399833    58399876     29089   
51  89d51m00s   9999966          34   -59453444    59453478     26180   
52  89d52m00s   9999973          27   -60631279    60631306     23271   
53  89d53m00s   9999979          21   -61966597    61966618     20362 